src/v/cluster/log_eviction_stm.cc

```diff
@@ -50,69 +52,107 @@ ss::future<> log_eviction_stm::start() {
+    _queue_mutex.broken();
+    _queue.abort(std::make_exception_ptr(ss::abort_requested_exception()));
     co_await persisted_stm::stop();
 }
 
+ss::future<> log_eviction_stm::enqueue_eviction_event(
+  model::offset offset, retry_event_until_success wait) {
+    co_await _queue_mutex.get_units();
+    co_await _queue.push_eventually(eviction_event{
+      .prefix_truncate_offset = offset, .wait_for_success = wait});
+}
+
 ss::future<> log_eviction_stm::write_raft_snapshots_in_background() {
+    static constexpr auto retry_backoff_time = 5s;
     /// This method is executed as a background fiber and it attempts to write
     /// snapshots as close to effective_start_offset as possible.
     auto gh = _gate.hold();
-    bool wait_with_timeout = false;
+
     while (!_as.abort_requested()) {
         /// This background fiber can be woken-up via apply() when special
         /// batches are processed or by the storage layer when local
         /// eviction is triggered.
         try {
-            static const auto eviction_event_wait_time = 5s;
-            if (wait_with_timeout) {
-                /// If after call to \ref do_write_snapshot() has not
-                /// truncated up until the desired point, it will be OK to
-                /// iterate again after some time to retry. Maybe
-                /// max_collectible_offset has moved forward.
-                co_await _reap_condition.wait(eviction_event_wait_time);
-            } else {
-                /// Last truncation had completed with success, there is no need
-                /// to wait for any other condition then for notify() to be
-                /// called.
-                co_await _reap_condition.wait();
-            }
-        } catch (const ss::condition_variable_timed_out&) {
-            /// There is still more data to truncate
+            co_await _queue.not_empty();
+        } catch (const ss::abort_requested_exception&) {
+            break;
         } catch (const ss::broken_condition_variable&) {
             /// stop() has been called
             break;
+        } catch (const ss::broken_semaphore&) {
+            /// stop() has been called
+            break;
         }
+
+        auto& event = _queue.front();
+        vlog(
+          _log.trace,
+          "processing prefix truncation event with offset: {}, waiting for "
+          "truncate: {}",
+          event.prefix_truncate_offset,
+          event.wait_for_success);
+
         auto evict_until = std::max(
-          _delete_records_eviction_offset, _storage_eviction_offset);
+          _delete_records_eviction_offset, event.prefix_truncate_offset);
         auto index_lb = _raft->log().index_lower_bound(evict_until);
         if (!index_lb) {
-            wait_with_timeout = false;
+            vlog(
+              _log.warn,
+              "unable to find index lower bound for {}",
+              evict_until);
+            if (
+              !event.wait_for_success
```

Comment:
here `event` is accessed after potentially being popped from the queue, dead reference

Reply:
This actually was the reason I was experiencing the heap use-after-free, copying the front item when calling `front` resolved the issue on my end